src/utils/havoc.js

```diff
@@ -411,7 +411,7 @@ class ObjectValueHavocingVisitor {
 
 function ensureFrozenValue(realm, value, loc) {
   // TODO: This should really check if it is recursively immutability.
-  if (value instanceof ObjectValue && !TestIntegrityLevel(realm, value, "frozen")) {
+  if (value instanceof ObjectValue && (!TestIntegrityLevel(realm, value, "frozen") && !value.isFinalObject())) {
```

Comment:
Nit: I don't see why add a paren here. It made me think you're negating the whole expression at first.